skorch/helper.py

```diff
@@ -131,6 +132,82 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
 
+def _unpack_first(x):
+    if isinstance(x, tuple):
+        return x[0]
+    return x
+
+
+class SliceDatasetX:
+    """Helper class that wraps a torch dataset to make it work with
+    sklearn.
+
+    Sometimes, sklearn will touch the input data, e.g. when splitting
+    the data for a grid search. This will fail when the input data is
+    a torch dataset. To prevent this, use this wrapper class for your
+    dataset.
+
+    Note that this class will only return the X value (i.e. the first
+    value returned by indexing the original dataset). Sklearn, and
+    hence skorch, always require 2 values, X and y. Therefore, you
+    still need to provide the y data separately.
+
+    Examples
+    --------
+    >>> X = MyCustomDataset()
+    >>> search = GridSearchCV(net, params, ...)
+    >>> search.fit(X, y)  # raises error
+    >>> ds = SliceDatasetX(X)  # or Xs = SliceDict(**X)
+    >>> search.fit(ds, y)  # works
+
+    Parameters
+    ----------
+    dataset : torch.utils.data.Dataset
+      A valid torch dataset.
+
+    collate_fn : callable (default=torch.utils.data.dataloader.default_collate)
+      A function that merges a list of samples to form a
+      mini-batch. This is typically the same function that torch's
+      DataLoader class uses to collate batches.
+
+    """
+    def __init__(self, dataset, collate_fn=default_collate):
+        self.dataset = dataset
+        self.collate_fn = collate_fn
+
+        self._indices = list(range(len(self.dataset)))
+
+    def __len__(self):
+        return len(self.dataset)
+
+    @property
+    def shape(self):
+        return (len(self),)
+
+    def transform(self, X):
+        return X
+
+    def __getitem__(self, i):
+        if isinstance(i, (int, np.integer)):
+            Xi = _unpack_first(self.dataset[i])
+            return self.transform(Xi)
```

Comment:
Currently, `self.transform` returns `X`, do we intend on letting users to subclass `SliceDatasetX` to provide their own transform?

Reply:
That's the intent. I will document it better.